SwiftVoiceQuickstart/ViewController.swift

```diff
@@ -453,6 +454,82 @@ extension ViewController: TVOCallDelegate {
         placeCallButton.setTitle("Call", for: .normal)
     }
     
+    func call(_ call: TVOCall, didReceiveQualityWarnings currentWarnings: Set<NSNumber>, previousWarnings: Set<NSNumber>) {
+        /**
+        * currentWarnings: existing quality warnings that have not been cleared yet
+        * previousWarnings: last set of warnings prior to receiving this callback
+        *
+        * Example:
+        *   - currentWarnings: { A, B }
+        *   - previousWarnings: { B, C }
+        *   - intersection: { B }
+        *
+        * Newly raised warnings = currentWarnings - intersection = { A }
+        * Newly cleared warnings = previousWarnings - intersection = { C }
+        */
+        var warningsIntersection: Set<NSNumber> = currentWarnings
+        warningsIntersection = warningsIntersection.intersection(previousWarnings)
+        
+        var newWarnings: Set<NSNumber> = currentWarnings
+        newWarnings.subtract(warningsIntersection)
+        if newWarnings.count > 0 {
+            qualityWarningsUpdatePopup(newWarnings, isCleared: false)
+        }
+        
+        var clearedWarnings: Set<NSNumber> = previousWarnings
+        clearedWarnings.subtract(warningsIntersection)
+        if clearedWarnings.count > 0 {
+            qualityWarningsUpdatePopup(clearedWarnings, isCleared: true)
+        }
+    }
+    
+    func qualityWarningsUpdatePopup(_ warnings: Set<NSNumber>, isCleared: Bool) {
+        var popupMessage: String! = "Warnings detected:"
+        if isCleared {
+            popupMessage = "Warnings cleared:"
+        }
+        
+        for warning in warnings {
+            if let warningName = warningString(TVOCallQualityWarning(rawValue: warning.uintValue)!) {
+                popupMessage = popupMessage + " " + warningName
+            }
+        }
```

Comment:
I think this can be simplified with `compactMap` and `joined`. Maybe even end up being a single line.

https://developer.apple.com/documentation/swift/sequence/2950916-compactmap
https://www.hackingwithswift.com/example-code/arrays/how-to-join-an-array-of-strings-into-a-single-string